### PR TITLE
59 ac clustering bu

### PR DESCRIPTION
**Pull request by annacanal**

### CODE REVIEW

**Comment by giovannivolpe** [comment]
@annacanal 
The code can be made more streamlined if Clustering is a child of Triangles (similarly to Degree and DegreeAv)

Please, check this with @egolol and @mitemijalkov and send it back to me once revised.

File: braph/graph/measures/Clustering.m [hunk 1]
```diff
@@ -0,0 +1,82 @@
+classdef Clustering < Measure
```
Comment (giovannivolpe) [inline braph/graph/measures/Clustering.m]: Clustering < Triangles

File: braph/graph/measures/Clustering.m [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+    methods
+        function m = Clustering (g, varargin)
+            m = m@Measure(g, varargin{:});
+        end
+    end
+    methods (Access=protected)
+        function clustering_coef = calculate(m)
+            g = m.getGraph();
```
Comment (giovannivolpe) [inline braph/graph/measures/Clustering.m]: Calculate triangles by calling the parent class at the beginning (advantage that it doesn't need to be saved).

            g = m.getGraph();
            
            if g.is_measure_calculated('Triangles')
                triangles = g.getMeasureValue('Triangles');
            else
                triangles = calculate@Triangles(m);
            end

File: braph/graph/measures/Clustering.m [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+            A = g.getA();
+            if isa(g, 'GraphBU') || isa(g, 'GraphWU')
+                triangles = g.getMeasureValue('Triangles');
+                degree = g.getMeasureValue('Degree');
+                clustering_coef = 2*triangles./(degree.*(degree-1));
+            elseif isa(g, 'GraphBD') || isa(g, 'GraphWD')
+                inDegree = g.getMeasureValue('InDegree');
```
Comment (giovannivolpe) [inline braph/graph/measures/Clustering.m]: call the variables In_degree and out_degree

for calculating the degrees use, e.g. (advantage not getting saved, less memory, also less code):
settings = g.getSettings();
degree = Degree(g, settings{:}).getValue();

File: braph/graph/measures/Clustering.m [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+    methods (Static)
+        function measure_class = getClass()
+            measure_class = 'Clustering';
+        end
+        function name = getName()
+            name = 'Clustering';
+        end
+        function description = getDescription()
+            description = [ ...
+                'The Clustering Coefficient of a node is ' ...
```
Comment (giovannivolpe) [inline braph/graph/measures/Clustering.m]: don't use capital for measure name in description

File: braph/graph/measures/test_Clustering.m [hunk 5]
```diff
@@ -0,0 +1,129 @@
+% test Clustering
```
Comment (giovannivolpe) [inline braph/graph/measures/test_Clustering.m]: @annacanal add spaces for readability
Check how to with @mitemijalkov @egolol

**Comment by giovannivolpe** [comment]
@annacanal @mitemijalkov 
When I run test_braph2, there is an error in Test 8.

File: braph/graph/measures/test_Clustering.m [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+A_test = rand(randi(10));
+A_BU = [
```
Comment (giovannivolpe) [inline braph/graph/measures/test_Clustering.m]: The test function should check all rules independently at least with the explicit matrices.

**Comment by annacanal** [comment]
I've made all the changes required.
But I still don't understand why the test 8, now test 10, gives an error, since when I check the 2 results being compared they are both exactly the same. Sometimes the test works, and sometimes it fails, although both time I can see both variables with the same value
Any idea? @mitemijalkov @egolol?  Could someone check?

value_braph2 =

   1.928824521265286
   1.250006569672336
   1.848974171928520
   1.796936727704220
   1.913945877672339
   1.568329104018380
   2.055750463109955
   2.459934110723147
   1.386475094699737
   1.898969373263950


value_std =

   1.928824521265286
   1.250006569672336
   1.848974171928520
   1.796936727704220
   1.913945877672339
   1.568329104018380
   2.055750463109955
   2.459934110723146
   1.386475094699737
   1.898969373263949

Error using test_Clustering (line 118)
Clustering [all] is not being calculated correctly for GraphWD

**Comment by annacanal** [comment]
All changes made and all test working.
@egolol @mitemijalkov

File: braph/graph/measures/test_ClusteringAv.m [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+    'ClusteringAv() should default to ClusteringAv(''DirectedTrianglesRule'', ''cycle'') for GraphBD')
+
+%% Test 3: Comparison with known BD graph - cycle
+g = GraphBD(A_BD);
+clustering_av_1 = ClusteringAv(g, 'DirectedTrianglesRule', 'cycle').getValue();
+clustering_av_2 = mean(clustering_BD_cycle);
+assert(isequal(clustering_av_1, clustering_av_2), ...
+    'BRAPH:Clustering:Bug', ...
+    'ClusteringAv(''DirectedTrianglesRule'', ''in'') is not being calculated correctly for GraphBD')
+
```
Comment (giovannivolpe) [inline braph/graph/measures/test_ClusteringAv.m]: @annacanal 

Please, update all tests according to what I did to tests 1-3.
in particular, ay attention to indentation and naming conventions.

File: braph/graph/measures/test_Clustering.m [hunk 8]
```diff
@@ -0,0 +1,173 @@
+% test Clustering
```
Comment (giovannivolpe) [inline braph/graph/measures/test_Clustering.m]: @annacanal 

Please, edit this like test_ClusteringAv (do that first).

File: braph/graph/measures/Clustering.m [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+        function measure_class = getClass()
+            measure_class = 'Clustering';
+        end
+        function name = getName()
+            name = 'Clustering';
+        end
+        function description = getDescription()
+            description = [ ...
+                'The clustering coefficient of a node is ' ...
+                'the fraction of triangles present around a node.' ...
```
Comment (egolol) [inline braph/graph/measures/Clustering.m]: 'the fraction of triangles present around a node. '

File: braph/graph/measures/Clustering.m [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+                'The clustering coefficient is calculated as the ratio between' ...
```
Comment (egolol) [inline braph/graph/measures/Clustering.m]: 'The clustering coefficient is calculated as the ratio between '

File: braph/graph/measures/Clustering.m [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+                'the number of triangles present around a node and' ...
```
Comment (egolol) [inline braph/graph/measures/Clustering.m]: 'the number of triangles present around a node and '

File: braph/graph/measures/Clustering.m [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+                'the maximum number of triangles that could possibly' ...
```
Comment (egolol) [inline braph/graph/measures/Clustering.m]: 'the maximum number of triangles that could possibly '

**Comment by egolol** [comment]
Very good code. Just I think there's a little trouble with spaces.

**Pull request merged**